File: rust/src-tauri/src/polling.rs
// Refresh scheduling, 1:1 port of QuotaService.SafeRefresh/Reschedule (UI-SPEC 7.5):
// first refresh 2s after start, period = max(1, RefreshMinutes), failure keeps
// last-known-good data and retries fast after 30s.

use crate::quota::{self, QuotaResult};
use crate::state::AppState;
use crate::tray;
use std::time::Duration;
use tauri::{AppHandle, Emitter, Manager};

pub async fn safe_refresh(app: &AppHandle) -> QuotaResult {
    let mut r = quota::fetch().await;
    let st = app.state::<AppState>();
    if r.error.is_some() {
        let last = st.last_quota.read().unwrap().clone();
        if let Some(l) = &last {
            r.fill_missing_from(l);
        }
    }
    *st.last_quota.write().unwrap() = Some(r.clone());
    let _ = app.emit("quota-updated", &r);
    tray::update_tooltip(app);
    // Every refresh (scheduled, manual, or hover) moves the next scheduled
    // tick: 30s after a failure, one period after a success (SPEC 7.5 step 3,
    // mirrors _timer.Change in QuotaService.SafeRefresh)
    let mins = st.settings.read().unwrap().refresh_minutes.max(1) as u64;
    let delay = if r.error.is_some() {
        Duration::from_secs(30)
    } else {
        Duration::from_secs(mins * 60)
    };
    *st.retime_delay.lock().unwrap() = Some(delay);
    st.retime.notify_one();
    r
}

/// Polling loop. `reschedule` notify restarts the cycle with the 2s first
/// delay (mirrors QuotaService.Reschedule's timer.Change(2s, period)).
pub async fn run(app: AppHandle) {
    let mut next = Duration::from_secs(2);
    loop {
        let reschedule = app.state::<AppState>().reschedule.clone();
        let retime = app.state::<AppState>().retime.clone();
        tokio::select! {
            _ = tokio::time::sleep(next) => {}
            _ = reschedule.notified() => {
                next = Duration::from_secs(2);
                continue;
            }
            _ = retime.notified() => {
                let hint = app.state::<AppState>().retime_delay.lock().unwrap().take();
                if let Some(d) = hint {
                    next = d;
                }
                continue;
            }
        }
        let r = safe_refresh(&app).await;
        next = if r.error.is_some() {
            Duration::from_secs(30) // fast retry after failure
        } else {
            let mins = app
                .state::<AppState>()
                .settings
                .read()
                .unwrap()
                .refresh_minutes
                .max(1) as u64;
            Duration::from_secs(mins * 60)
        };
    }
}

File: qt/src/polling.cpp
#include "polling.h"

#include "app.h"
#include "async_util.h"
#include "quota.h"
#include "state.h"
#include "trayicon.h"
#include "windows/panelwindow.h"

#include <QTimer>

Poller::Poller(App *app) : QObject(app), m_app(app)
{
    m_timer = new QTimer(this);
    m_timer->setSingleShot(true);
    connect(m_timer, &QTimer::timeout, this, &Poller::safeRefresh);
}

void Poller::start()
{
    m_timer->start(2000); // first refresh 2s after launch (SPEC 16.5)
}

void Poller::reschedule()
{
    m_timer->start(2000); // mirrors timer.Change(2s, period) on settings save
}

qint64 Poller::successDelayMs() const
{
    const qint64 mins = std::max<qint64>(1, m_app->state().settings.refreshMinutes);
    return mins * 60 * 1000;
}

void Poller::safeRefresh()
{
    // A fetch already in flight is left alone (the Rust edition would spawn a
    // parallel one; the result and retiming are identical either way).
    if (m_inFlight)
        return;
    m_inFlight = true;
    runAsync<QuotaResult>(this, [] { return quota::fetch(); },
                          [this](QuotaResult &&r) { onFetched(std::move(r)); });
}

void Poller::onFetched(QuotaResult &&r)
{
    m_inFlight = false;
    AppState &st = m_app->state();
    if (r.error && st.lastQuota)
        r.fillMissingFrom(*st.lastQuota);
    st.lastQuota = r;
    m_app->panelWindow()->renderQuota(st.lastQuota);
    m_app->trayIcon()->updateTooltip();
    // Every refresh (scheduled, manual, or hover) moves the next scheduled
    // tick: 30s after a failure, one period after a success (SPEC 16.5 step 3)
    m_timer->start(r.error ? 30000 : successDelayMs());
}
